[asan] Speed up ASan ODR indicator-based checking

artempyanykh opened the pull request:
**Summary**:
When ASan checks for a potential ODR violation on a global it loops over a linked list of all globals to find those with the matching value of an indicator. With the default setting 'detect_odr_violation=1', ASan doesn't report violations on same-size globals but it still has to traverse the list. For larger binaries with a ton of shared libs and globals (and a non-trivial volume of same-sized duplicates) this gets extremely expensive.

This patch adds an indicator indexed (multi-)map of globals to speed up the search.

> Note: asan used to use a map to store globals a while ago which was replaced with a list when the codebase [moved off of STL](https://github.com/llvm/llvm-project/commit/e4bada2c946e5399fc37bd67421de01c0047ad38).

Internally we see many examples where ODR checking takes *seconds* (even double digits). With this patch it's practically free and `__asan_register_globals` doesn't show up prominently in the perf profile anymore.

There are several high-level questions:
1. I understand that the intent is that we hit the slow path rarely, ideally once before the process dies with an error. But in practice we hit the slow path a lot. It feels reasonable to keep the amount of work bounded even in the worst case, even if it requires a bit of extra memory. But if not, it'd be great to learn about the tradeoffs.
2. Poisoning based ODR checking remains on the slow path. Internally we build everything with `-fsanitize-address-use-odr-indicator` so I'm not sure if poisoning-based check would exhibit the same behavior (looking at the code, the shape looks very similar, so it might?).
3. Globals with an ODR indicator of `-1` need to be skipped for the purposes of ODR checking (cf. https://github.com/llvm/llvm-project/commit/a257639a6935a2c63377784c5c9c3b73864a2582). But they are still getting added to the list of globals and hence take up space and slow down the iteration over the list of globals. It would be a good saving if we could avoid adding them to the globals list.
4. Any reason to use a linked list instead of e.g. a vector to store globals?

**Test Plan**:

* `cmake --build build --target check-asan` looks good
* Perf-wise things look good when linking against this version of compiler-rt.

artempyanykh committed:
[asan] Speed up ASan ODR indicator-based checking

When ASan checks for ODR violation on a global it loops over a linked list of all globals to find
those with the matching value of an indicator. For larger binaries with a ton of globals this gets
extremely expensive: O(N^2) iterations + potentially poor memory locality from the linked list.

This patch adds an indicator indexed (multi-)map of globals to speed up the search. With a bit of
extra memory this massively speeds up indicator based ODR checks.

Poisoning based ODR checking remains on the slow path.

vitalybuka commented:
Thanks for the patch!
We aware of performance issues in this check, but had not time to make it a priority.

MaskRay commented on compiler-rt/lib/asan/asan_globals.cpp:
```diff
@@ -147,14 +163,20 @@ static void CheckODRViolationViaIndicator(const Global *g) {
+  auto *relevant_globals_lookup =
+      map_of_globals_by_indicator.find(g->odr_indicator);
+  if (!relevant_globals_lookup) {
+    return;
```
remove braces s for single-line simplebody

MaskRay commented on compiler-rt/lib/asan/asan_globals.cpp:
```diff
@@ -147,14 +163,20 @@ static void CheckODRViolationViaIndicator(const Global *g) {
     *odr_indicator = REGISTERED;
     return;
   }
+  // Fetch globals with the same ODR indicator
```
Comments end with a period.

artempyanykh committed:
amend! [asan] Speed up ASan ODR indicator-based checking

[asan] Speed up ASan ODR indicator-based checking

When ASan checks for a potential ODR violation on a global it loops over
a linked list of all globals to find those with the matching value of an
indicator. With the default setting 'detect_odr_violation=1', ASan
doesn't report violations on same-size globals but it still has to
traverse the list.

For larger binaries with a ton of shared libs and globals (and a
non-trivial volume of same-sized duplicates) this gets extremely
expensive.

This patch adds an indicator indexed (multi-)map of globals to speed up
the search.

artempyanykh committed:
fixup! amend! [asan] Speed up ASan ODR indicator-based checking

artempyanykh commented:
Thanks for the context @vitalybuka! This has become a bit of a bottleneck for us, hence this patch.
And thanks for review comments @MaskRay!

Looking forward to hear what you think.

vitalybuka commented on compiler-rt/lib/asan/asan_globals.cpp:
```diff
@@ -225,10 +248,12 @@ static void RegisterGlobal(const Global *g) {
+
+  AddGlobalToList(list_of_all_globals, g);
+
+  if (UseODRIndicator(g) && g->odr_indicator != UINTPTR_MAX)
```
I'd prefer we do this in CheckODRViolationViaIndicator, 
so instead of 2 lookups: find + [], 
just one one [] there.

artempyanykh replied:
I can do that. The only concern is that `Check` will be more like `CheckAndAdd` rather than just doing checking. OTOH we won't need to repeat this guard.

vitalybuka replied:
I don't have problem with the current name, we do the `check`, and adding into the map is just `check` implementation detail.

artempyanykh replied:
@vitalybuka I moved things around. LMK if now looks good.

vitalybuka commented on compiler-rt/lib/asan/asan_globals.cpp:
```diff
@@ -147,14 +163,21 @@ static void CheckODRViolationViaIndicator(const Global *g) {
-    if (g->odr_indicator == l->g->odr_indicator &&
-        (flags()->detect_odr_violation >= 2 || g->size != l->g->size) &&
+  for (ListOfGlobals *l = relevant_globals; l; l = l->next) {
+    if ((flags()->detect_odr_violation >= 2 || g->size != l->g->size) &&
```
Do we need full list?
why not to keep just the first global, and raise an error on the first mismatch?

artempyanykh replied:
When `halt_on_error=1`, the first one's enough. If `halt_on_error=0`, my understanding is that keeping just the first one may make it report less violations. If keeping the existing behavior is not necessary, I can make it keep just the first global. We always halt on error, so for us just the first one would be enough. With that said, even when keeping the full list the perf was great compared to the current baseline.

LMK what you prefer.

vitalybuka replied:
Taking into account we may still need unregistered

vitalybuka commented on compiler-rt/lib/asan/asan_globals.cpp:
```diff
@@ -147,14 +163,21 @@ static void CheckODRViolationViaIndicator(const Global *g) {
   // If *odr_indicator is DEFINED, some module have already registered
   // externally visible symbol with the same name. This is an ODR violation.
-  for (ListOfGlobals *l = list_of_all_globals; l; l = l->next) {
-    if (g->odr_indicator == l->g->odr_indicator &&
```
the patch does not take into account that global can be unregistered.
As is it will stay in map with invalid indicator.

artempyanykh replied:
Good point! This is in line with the current behavior where it also stays in the list with an `UNREGISTERED` indicator. If we're ok keeping just the first global in the map, then removing from the map when unregistering is easy. If we need the whole list, removing is also not hard and should be still fast enough.

LMK if it's important to have removal from the map as part of this patch (or maybe as a patch on top).

vitalybuka replied:
Instead of removal checking that indicator matches is enough

we can't keep just first one with UNREGISTERED, if we unregister it we don't know if there are other with same indicator are still registered

artempyanykh replied:
> Instead of removal checking that indicators still match is enough

@vitalybuka could you expand on what you mean by 'indicators still match'?

Looking at the code for `UnregisterGlobal`, we update the value of whatever the indicator points to to `UNREGISTERED` (which affects all instances that share the same indicator) but the indicator address is unchanged.
https://github.com/llvm/llvm-project/blob/99b2444c7742b2c3a955660c75a95e063c03fc74/compiler-rt/lib/asan/asan_globals.cpp#L283-L284

I didn't find any tests around 'unregistering' so I added one in https://github.com/llvm/llvm-project/pull/100923/commits/01847d1dd72fb078d26f38973391f37d2815ec3c to demonstrate the behavior (it's the same on main and in this branch).

vitalybuka replied:
LGTM
I was confused by `odr_indicator` vs `*odr_indicator`.

artempyanykh committed:
fixup! fixup! amend! [asan] Speed up ASan ODR indicator-based checking

artempyanykh committed:
[asan] Add a test that captures the ODR checker behavior when globals get unregistered

The test demonstrates a FN due to indicator value being shared among all globals with the same
indicator address.

Moving dlopen/dlclose around and/or adding more shared libs into the mix can show more edge cases
but this is out of the scope of this commit.

vitalybuka commented on compiler-rt/lib/asan/asan_globals.cpp:
```diff
@@ -35,8 +36,11 @@ struct ListOfGlobals {
   ListOfGlobals *next;
 };
 
+typedef DenseMap<uptr, ListOfGlobals *> MapOfGlobals;
```
typedef DenseMap<uptr, IntrusiveList<ListOfGlobals>> MapOfGlobals;

artempyanykh replied:
Replaced here and in the main list for consistency https://github.com/llvm/llvm-project/pull/100923/commits/151257d3d68e932b1b13f45f3b204a2400637ce8

vitalybuka replied:
It would be nice to have such small refactoring out of real patches.
Assuming we are in different time zones, to avoid unnecessary iterations, I will extract land it separately.

artempyanykh replied:
Thanks @vitalybuka, appreciate it! 

P.S. If there's a guide on how to do "stacked" PRs on Github or otherwise what's the blessed workflow for stacks of changes that'd be very helpful (not much about this in https://llvm.org/docs/GitHub.html). I'd do a stack of diffs in Phabricator but here it's different.

artempyanykh replied:
Thanks for the pointers and the explanation @vitalybuka!

artempyanykh committed:
[asan][odr] Use IntrusiveList for a ListOfGlobals

vitalybuka committed:
Revert "[NFC][asan][odr] Use IntrusiveList for a ListOfGlobals"

This reverts commit 2a5f7e58d71c394b241fdd1d905041ad0537acab.

vitalybuka committed:
rebase

vitalybuka reviewed (approved):
LGTM. Will land in 30min or so.

artempyanykh commented:
Thanks for the review @vitalybuka!

vitalybuka commented:
BTW. 
My concern is slowness of 
__asan_after_dynamic_init/__asan_before_dynamic_init
They are 10x slower then unpatched `register`.
However it's not enabled by default, so maybe it's unused in your env.

I have patches for that and will CC you.

artempyanykh commented:
Not sure I've seen those prominently in profiles. Will double check. Thanks for the heads up @vitalybuka!

artempyanykh commented:
Just for my own curiosity, but @vitalybuka where/how DenseMap's destructor was causing problems? https://github.com/llvm/llvm-project/commit/54c9404976118b0db5b20448e976db7a8c5b7073

vitalybuka commented:
> Just for my own curiosity, but @vitalybuka where/how DenseMap's destructor was causing problems? [54c9404](https://github.com/llvm/llvm-project/commit/54c9404976118b0db5b20448e976db7a8c5b7073)

I don't a particular case, we avoid global destructors in general.

vitalybuka commented:
and constructors
